source/k64f_eth_nanostack_port.c

```diff
@@ -501,236 +361,275 @@ static int8_t arm_eth_phy_k64f_address_write(phy_address_type_e address_type, ui
-        last_known_status = LINK_STATUS_UP;
-        ethernet_iface_ptr->isInitialized = true;
-        phy_link_wrapper_create();
-        return 0;
+        eth_if_unlock();
+        driver_readiness_status_callback(link, eth_interface_id);
+        tr_info ("Ethernet cable %s.", link ? "connected" : "unplugged");
+    } else {
+        eth_if_unlock();
     }
 
-    return -1;
+    eventOS_timeout_ms(PHY_LinkStatus_Task, 500, NULL);
 }
 
-void k64f_eth_phy_link_poll()
+
+/** \brief  Initialization function for Ethernet module.
+ *
+ *      This function initializes the ethernet module using default
+ *      configuration. Link speed and duplex is auto-negotiated. It also sets up
+ *      the buffer descriptors. If MAC address is setup later by the stack using
+ *      k64f_eth_address_set() routine, that will just overwrite the MAC directly
+ *      in Hardware registers.
+ *
+ *  \param[in] mac_addr Pointer to MAC address.
+ *
+ */
+static void k64f_eth_initialize(uint8_t *mac_addr)
 {
-    bool link_status = false;
-    enet_dev_if_t *ethernet_iface_ptr =
-            &ethernet_iface[BOARD_DEBUG_ENET_INSTANCE];
-
-    /* First check, if the cable is connected or not.*/
-    phy_get_link_status(ethernet_iface_ptr, &link_status);
-
-    if (!link_status && last_known_status==LINK_STATUS_UP) {
-        driver_readiness_status_callback(0, eth_interface_id);
-        last_known_status = LINK_STATUS_DOWN;
-        ethernet_iface_ptr->isInitialized = false;
-        tr_debug("Ethernet phy link status = down");
-        tr_info("Ethernet Cable is not connected.");
+    uint32_t sysClock;
+    phy_speed_t phy_speed;
+    phy_duplex_t phy_duplex;
+    uint32_t phyAddr = 0;
+
+    enet_config_t config;
+
+
+    /* Allocate TX buffer descriptors */
+    void *memptr = MEM_ALLOC(
+            sizeof(enet_tx_bd_struct_t) * ENET_TX_RING_LEN + ENET_BUFF_ALIGNMENT);
+    if (!memptr) {
+        tr_error("Could not allocate memory for TX Buf Descriptors.");
+        return;
+    }
+    base_DS.tx_buf_desc_start_addr =
+            (volatile enet_tx_bd_struct_t *) ENET_ALIGN(memptr, ENET_BUFF_ALIGNMENT);
+    memset((enet_tx_bd_struct_t *) base_DS.tx_buf_desc_start_addr, 0,
+           sizeof(enet_tx_bd_struct_t) * ENET_TX_RING_LEN);
+
+    /* Allocate RX buffer descriptors */
+    memptr = MEM_ALLOC( sizeof(enet_rx_bd_struct_t) * ENET_RX_RING_LEN + ENET_BUFF_ALIGNMENT);
+    if (!memptr) {
+        tr_error("Could not allocate memory for RX Buf Descriptors");
+        return;
+    }
+    base_DS.rx_buf_desc_start_addr = (volatile enet_rx_bd_struct_t *) ENET_ALIGN(memptr,
+                                                        ENET_BUFF_ALIGNMENT);
+    memset((enet_rx_bd_struct_t *) base_DS.rx_buf_desc_start_addr, 0,
+           sizeof(enet_rx_bd_struct_t) * ENET_RX_RING_LEN);
+
+    /* Allocating buffers for each RX Buffer Descriptor */
+    for (int i = 0; i < ENET_RX_RING_LEN; i++) {
+        base_DS.rx_buf[i] = MEM_ALLOC(
```

Comment:
It would be a tad more efficient to claim all RX buffers in one block, as is being done for the descriptors.

Note that the comment says "descriptor" wrongly.